src/client/clientmap.cpp

```diff
@@ -180,17 +204,420 @@ void ClientMap::getBlocksInViewRange(v3s16 cam_pos_nodes,
+			u8 ignore_inner_sides = (look.X == 0 ? 3 : 0) |
+				(look.Y == 0 ? 12 : 0) |
+				(look.Z == 0 ? 48 : 0);
+
+			// get the mask for the sides that are relevant based on the direction
+			u8 near_inner_sides = (look.X > 0 ? 1 : 2) |
+					(look.Y > 0 ? 4 : 8) |
+					(look.Z > 0 ? 16 : 32);
+			
+			// This bitset is +Z-Z+Y-Y+X-X (See MapBlockMesh), and axis is XYZ.
+			u8 transparent_sides = (occlusion_culling_enabled && block) ? ~block->solid_sides : 0x3F;
+
+			// compress block transparent sides to ZYX mask of see-through axes
+			u8 near_transparency = ((transparent_sides & near_inner_sides) | ignore_inner_sides) & 0x3F;
+
+			near_transparency = (near_transparency | (near_transparency >> 1)) & 0x15;
+			near_transparency = (near_transparency & 1) |
+					((near_transparency >> 1) & 2) |
+					((near_transparency >> 2) & 4);
+
+			// combine with known visible sides that matter
+			near_transparency &= visible_outer_sides;
+
+			// The rule for any far side to be visible:
+			// * Any of the adjacent near sides is transparent (different axes)
+			// * The opposite near side (same axis) transparent, and is the dominant axis of the look vector
+
+			// dominant axis flag
+			u8 dominant_axis = (abs(look.X) > abs(look.Y) && abs(look.X) > abs(look.Z)) |
+						((abs(look.Y) > abs(look.Z) && abs(look.Y) > abs(look.Z)) << 1) |
+						((abs(look.Z) > abs(look.X) && abs(look.Z) > abs(look.Y)) << 2);
+
+			// Queue next blocks for processing:
+			// - Examine "far" sides of the current blocks, i.e. never move towards the camera
+			// - Only traverse the sides that are not occluded
+			// - Only traverse the sides that are not opaque
+			// When queueing, mark the relevant side on the next block as 'visible'
+			for (s16 axis = 0; axis < 3; axis++) {
+
+				// Select a bit from transparent_sides
+				u8 transparency_mask = 1 << (2 * axis);
+
+				// axis flag
+				u8 my_side = 1 << axis;
+				u8 adjacent_sides = my_side ^ 0x07;
+
+				// Test the '-' direction of the axis
+				if (look[axis] <= 0 && block_coord[axis] > p_blocks_min[axis]) {
+					// far side is visible if adjacent near sides are transparent, or if opposite side on dominant axis is transparent
+					bool side_visible = ((near_transparency & adjacent_sides) | (near_transparency & my_side & dominant_axis)) != 0;
+					side_visible = side_visible && ((transparency_mask & transparent_sides) != 0);
+
+					v3s16 next_pos = block_coord;
+					next_pos[axis] -= 1;
+
+					// If a side is a see-through, mark the next block's side as visible, and queue
+					if (side_visible) {
+						auto &next_flags = blocks_seen.getChunk(next_pos).getBits(next_pos);
+						next_flags |= my_side;
+						blocks_to_consider.push(next_pos);
+					}
+					else {
+						sides_skipped++;
+					}
+				}
+
+				// Test the '+' direction of the axis
+				transparency_mask <<= 1;
+				if (look[axis] >= 0 && block_coord[axis] < p_blocks_max[axis]) {
```

Comment:
`p_blocks_max` is exclusive, I assume.
If `block_coord[axis] + 1 == p_blocks_max[axis]`, `next_pos` will be outside of range, and cause a buffer-overflow in `blocks_seen`.
```suggestion
				if (look[axis] >= 0 && block_coord[axis] + 1 < p_blocks_max[axis]) {
```

Reply:
It's inclusive and was inclusive in the previous algorithm. blocks_seen has a + 1 in constructor to avoid overflow.